### PR TITLE
DOC: clarify //pose/@relative_to

### PR DESCRIPTION
**Pull request by FirefoxMetzger**
Thanks to @azeey and @scpeters pointing me to relevant documentation my understanding of how the pose semantics are supposed to work has improved significantly. I thought I use this knowledge for some good and try to clarify the spec. Please double-check if I got it right.

### CODE REVIEW

**Comment by FirefoxMetzger** [comment]
RTR from my side. This morning I found some more docs on scoping behavior and when implicit frames are created. Once this is clear the current pose semantics become much more obvious. I tried to reflect that in the docs of the spec.

**Comment by FirefoxMetzger** [comment]
Actually, I am not sure if the current description mentions all the entities that generate implicit frames. For example, if I have a //sensor and in particular a //sensor[@type=camera] then there are one (or two) poses that can be specified, which may or may not use @relative_to.

Does this mean that their names, too, must be unique in the current scope? Can an object be placed relative to a //sensor? What about other elements like //visual, or //collision?

**Comment by FirefoxMetzger** [comment]
I think it's just a bad sample, but the CI pipeline really doesn't seem to stable for my PRs here ... it's been building for two days now.

> Does this mean that their names, too, must be unique in the current scope? Can an object be placed relative to a //sensor? What about other elements like //visual, or //collision?

Regarding my earlier question, I think I figured it out for //collision since the test file `model_frame_relative_to_joint.sdf` declares a collision element with the same name as a link and - as far as I can tell - this is expected to be a valid SDF. So for //collision and //visual, I think it shouldn't create a new frame and consequentially you can't place objects relative to them.

Still not sure about a //sensor.

@azeey Any thoughts on my doc update otherwise?

File: sdf/1.7/frame.sdf [hunk 1]
```diff
@@ -1,16 +1,31 @@
...
+
+      If missing, this frame is attached to the containing scope's frame. Within
+      a //world scope this is the implicit world frame, and within a //model
+      scope this is the implicit frame defined by the model's canonical link.
+
+      A frame moves jointly with the frame it is @attached_to. This is different
+      from //pose/@relative_to. @attached_to defines how the frame is attached
+      to the scope's frame, while //pose/@relative_to defines how the frame's
+      pose is represented numerically. As a result, following the chain of
+      @attached_to attributes must always lead to the scope's frame.
```
Comment (scpeters) [inline sdf/1.7/frame.sdf:28]: > following the chain of @attached_to attributes must always lead to the scope's frame.

based on the discussion above, I would rephrase as "following the chain of @attached_to attributes must always lead to a link, model, or world frame."

File: sdf/1.7/model.sdf [hunk 2]
```diff
@@ -3,7 +3,11 @@
   <description>The model element defines a complete robot or any other physical object.</description>
 
   <attribute name="name" type="string" default="__default__" required="1">
-    <description>A unique name for the model. This name must not match another model in the world.</description>
+    <description>
+      The name of the model and its implicit frame. This name must be unique
+      among all elements defining frames within the same scope, i.e., it must
+      not match another //model, //frame, or //link within the same scope.
```
Comment (scpeters) [inline sdf/1.7/model.sdf:9]: add `//joint` to the list

**Review by scpeters — COMMENTED** [review]
I just started looking through this and will resume later when I have a chance. I've made a few comments for now.

Thanks for improving the documentation here!

File: sdf/1.7/frame.sdf [hunk 3]
```diff
@@ -1,16 +1,31 @@
...
+      the name of any //frame, //model, or //link within the same scope may be
+      used.
+
+      If missing, this frame is attached to the containing scope's frame. Within
+      a //world scope this is the implicit world frame, and within a //model
+      scope this is the implicit frame defined by the model's canonical link.
+
+      A frame moves jointly with the frame it is @attached_to. This is different
+      from //pose/@relative_to. @attached_to defines how the frame is attached
+      to the scope's frame, while //pose/@relative_to defines how the frame's
```
[the remark this answers is not in the document]
Comment (FirefoxMetzger) [reply sdf/1.7/frame.sdf:26]: @scpeters My thinking here was that any frame-bearing element (//link, //joint, //model, ...) has itself a frame that is (implicitly) attached to the scope's implicit frame via it's //pose. This way, everything traces back to the scope's implicit frame.

That said, I can still change to your suggestion, if you think it is more clear.

File: sdf/1.7/frame.sdf (hunk 1 above)
Comment (FirefoxMetzger) [reply sdf/1.7/frame.sdf:28]: Will change this depending on the outcome of above discussion.

File: sdf/1.7/frame.sdf (hunk 3 above)
[the remark this answers is not in the document]
Comment (scpeters) [reply sdf/1.7/frame.sdf:26]: the initial configuration of each frame is defined by its //pose relative to the scope's implicit frame, but as time passes the frames can move relative to each other and the scope's implicit frame depending on the kinematic relationships encoded in the //joint tags. Any frames that are `@attached_to` each other have unchanging relative position and orientation. Tracing the `@attached_to` tags can lead to any link or model or world frame (not joint frames since they are automatically attached to their child link frame)

**Comment by scpeters** [comment]
> Actually, I am not sure if the current description mentions all the entities that generate implicit frames. For example, if I have a //sensor and in particular a //sensor[@type=camera] then there are one (or two) poses that can be specified, which may or may not use @relative_to.

The 1.7 proposal for pose frame semantics discusses explicit and implicit frames in [section 1.2](http://sdformat.org/tutorials?tut=pose_frame_semantics_proposal#1-2-explicit-vs-implicit-frames) and states that only link, joint, model and world have implicit frames

> Does this mean that their names, too, must be unique in the current scope?

the name uniqueness rule in SDFormat 1.7 (see [section 3.2 of the proposal](http://sdformat.org/tutorials?tut=pose_frame_semantics_proposal#3-2-unique-names-for-all-sibling-elements)) is rather broad, requiring unique names among all sibling elements, even if they don't have frames.

>  Can an object be placed relative to a //sensor? What about other elements like //visual, or //collision?

The name of a `//sensor`, `//collision`, or `//visual` cannot be used in a `//pose/@relative_to` or `//frame/@attached_to` attribute as they do not have implicit frames. These tags can use the `//pose/@relative_to` attribute to define their own pose, but it must use the name of an implicit frame.

**Comment by scpeters** [comment]
> > Does this mean that their names, too, must be unique in the current scope? Can an object be placed relative to a //sensor? What about other elements like //visual, or //collision?
> 
> Regarding my earlier question, I think I figured it out for //collision since the test file `model_frame_relative_to_joint.sdf` declares a collision element with the same name as a link and - as far as I can tell - this is expected to be a valid SDF. So for //collision and //visual, I think it shouldn't create a new frame and consequentially you can't place objects relative to them.

this is correct
 
> Still not sure about a //sensor.

it is similar to `//collsiion` and `//visual`; you can't place objects relative to a `//sensor`. You can, however, place a `//frame` where the sensor will be and then place the sensor relative to that `//frame`, so there is a different way to accomplish the desired posturing

File: sdf/1.7/frame.sdf (hunk 3 above)
[the remark this answers is not in the document]
Comment (FirefoxMetzger) [reply sdf/1.7/frame.sdf:26]: @scpeters Ah, I see where I was wrong this should indeed be clarified; thanks for the explanation. I'll see if I can add a sentence that would have resolved my confusion from the get-go.

File: sdf/1.7/pose.sdf [hunk 4]
```diff
@@ -4,7 +4,19 @@
...
+
+      If missing, this pose is expressed in the scope's implicit frame. This
+      frame is implicitly defined by the world or model that contains the
+      element that has its pose set by this pose element.
+
+      Regardless of the value of @relative_to, the element that has its pose
+      specified by this tag remains attached to its parent and moves with it.
+      Exceptions to this rule are //frame elements specifying @attached_to (see
+      //frame/@attached_to) and the implicit frames of //model elements setting
+      //model/@canonical_link.
```
Comment (scpeters) [inline sdf/1.7/pose.sdf:19]: other exceptions:

* //link frame moves on its own, not attached to its parent
* //joint frame is attached to its child link

would it be clearer to state that the pose element specifies the initial relative pose of an object and does not affect to which frame it is attached? Is this the best place to document attachment rules? I'm not sure

**Comment by scpeters** [comment]
Ok, I've finished looking through your improvements to the 1.7 documentation and just have one last comment there

There are some changes in 1.8, particularly with regard to the `::` delimiter token for referencing frames in nested scopes. This hasn't made its way into the documentation section yet, but it is described in the 1.8 composition proposal:

* http://sdformat.org/tutorials?tut=composition_proposal&ver=1.8&cat=pose_semantics_docs&#1-3-1-reserved-delimiter-token

File: sdf/1.7/pose.sdf (hunk 4 above)
Comment (FirefoxMetzger) [reply sdf/1.7/pose.sdf:19]: @scpeters Yep, it is better to not document attachment rules here; especially with those additional exceptions. I initially wrote this sentence with the wrong understanding of what //pose actually does and erroneously thought that it actually influences attachment, too. (And was very confused about how this works with links and joints.)

I'll revise the sentence based on your suggestion.

**Comment by FirefoxMetzger** [comment]
> There are some changes in 1.8, particularly with regard to the :: delimiter token for referencing frames in nested scopes. 

@scpeters Yep, I added docs for that in the v1.8 spec

```
New in v1.8: @relative_to may use frames of nested scopes within the same file. 
In this case, the frame is specified using `::` as delimiter to define the scope of the frame, e.g. `nested_model_A::nested_model_B::awesome_frame`.
```

File: sdf/1.7/frame.sdf [hunk 5]
```diff
@@ -1,16 +1,32 @@
 <!-- Frame -->
 <element name="frame" required="*">
-  <description>A frame of reference to which a pose is relative.</description>
+  <description>A frame of reference in which poses may be expressed.</description>
```
Comment (scpeters) [inline sdf/1.7/frame.sdf:3]: can you copy the change on this line to `sdf/1.8/frame.sdf`?

**Review by scpeters — COMMENTED** [review]
just one last change and then I think it will be good to go

File: sdf/1.7/frame.sdf (hunk 5 above)
Comment (FirefoxMetzger) [reply sdf/1.7/frame.sdf:3]: @scpeters Sure think ... done :)

File: sdf/1.7/pose.sdf [hunk 6]
```diff
@@ -4,7 +4,16 @@
 
   <attribute name="relative_to" type="string" default="" required="*">
     <description>
-      Name of frame relative to which the pose is applied.
+      If specified, this pose is expressed in the named frame. The named frame
+      must be declared both within the same file, and in the same scope
+      (world/model) as the element that has its pose specified by this tag.
+
+      If missing, this pose is expressed in the scope's implicit frame. This
+      frame is implicitly defined by the world or model that contains the
```
Comment (azeey) [inline sdf/1.7/pose.sdf:12]: The implicit frame used depends on the context of the pose. For `//visual` and `//collision`, their pose is relative to the parent `//link`, so saying the implicit frame is defined by the world or model is not accurate.

File: sdf/1.7/pose.sdf [hunk 7]
```diff
@@ -4,7 +4,16 @@
...
+      element that has its pose set by this pose element.
+
+      Note that @relative_to merely affects an element's initial position and
```
Comment (azeey) [inline sdf/1.7/pose.sdf:15]:
```suggestion
      Note that @relative_to merely affects an element's initial pose and
```

File: sdf/1.7/frame.sdf [hunk 8]
```diff
@@ -1,16 +1,32 @@
...
-      If a frame is specified, recursively following the attached_to attributes
-      of the specified frames must lead to the name of a link, a model, or the world frame.
+      If specified, this frame is attached to the specified frame. The specified
+      frame must be within the same scope and may be defined implicitly, i.e.,
+      the name of any //frame, //model, //joint, or //link within the same scope
+      may be used.
+
+      If missing, this frame is attached to the containing scope's frame. Within
+      a //world scope this is the implicit world frame, and within a //model
+      scope this is the implicit frame defined by the model's canonical link.
```
Comment (azeey) [inline sdf/1.7/frame.sdf:22]: The model's implicit frame (aka `__model__`) is different from the implicit frame of the canonical link. For example,
```xml
<model name='M'>
  <link name='L'>
    <pose>1 0 0   0 0 0</pose>
   </link>
</model>
```
The canonical link `L` is offset by pose `[1, 0, 0, 0, 0, 0]` from the model frame. So I think it would be better to say "...within a //model scope this is the implicit model frame".

File: sdf/1.8/pose.sdf [hunk 9]
```diff
@@ -4,7 +4,21 @@
...
+
+      If missing, this pose is expressed in the scope's implicit frame. This
+      frame is implicitly defined by the world or model that contains the
+      element that has its pose set by this pose element.
+
+      Note that @relative_to merely affects an element's initial position and
+      does not affect the element's dynamic movement thereafter.
+
+      New in v1.8: @relative_to may use frames of nested scopes within the same
+      file. In this case, the frame is specified using `::` as delimiter to
```
Comment (azeey) [inline sdf/1.8/pose.sdf:19]: Technically, a frame defined in another file can be referenced if that file (i.e. model) is nested via `//include`. Do we need to say "the same file"?

Comment (FirefoxMetzger) [reply sdf/1.8/pose.sdf:19]: No, we can remove it if it is possible to define @relative_to wrt. a frame inside an //included model. I added this part after reading `As a side effect, encapsulation is enforced: relative references are bounded according to the current file.`  here http://sdformat.org/tutorials?tut=composition_proposal&cat=pose_semantics_docs&#1-3-2-reference-types , but I guess I may have misunderstood it.

File: sdf/1.7/pose.sdf (hunk 6 above)
Comment (FirefoxMetzger) [reply sdf/1.7/pose.sdf:12]: @azeey Thanks, I didn't know that. Do you know if the default frame (by context) is documented somewhere? Perhaps we can link to it, or I can come up with a way of phrasing it consciesly here once I know the cases.

File: sdf/1.7/frame.sdf (hunk 8 above)
Comment (FirefoxMetzger) [reply sdf/1.7/frame.sdf:22]: Hm. That part is not quite clear to me yet. If I don't specify `@attached_to`  for a frame it will be attached to `__model__`  instead of being attached to the model's canonical link?

I totally get that the frame's pose will `@relative_to`  the `__model__` frame, but I was under the impression that `__model__` is only used for initial placement and not for dynamics. Isn't the canonical link the element that gets (dynamically) connected to the parent model/world? Then again, I might misunderstand the concept/purpose of the canonical link element.

@azeey I will update it as per your suggestion. At the same time, it would help me if you could comment on what I just said so that I can figure out what I am confusing.

File: sdf/1.7/pose.sdf (hunk 6 above)
Comment (azeey) [reply sdf/1.7/pose.sdf:12]: Yes, it's documented in http://sdformat.org/tutorials?tut=pose_frame_semantics&ver=1.5&cat=specification&#parent-frames-in-sdf-1-4. The later versions of the SDFormat spec simply keep that behavior when `relative_to` is not specified.

File: sdf/1.7/frame.sdf (hunk 8 above)
Comment (azeey) [reply sdf/1.7/frame.sdf]: > Hm. That part is not quite clear to me yet. If I don't specify @attached_to for a frame it will be attached to __model__ instead of being attached to the model's canonical link?

Yes, but note that the `__model__` frame in turn is attached to the canonical link.

> I totally get that the frame's pose will @relative_to the __model__ frame, but I was under the impression that __model__ is only used for initial placement and not for dynamics.

No, the `__model__` frame is the implicit frame of the model, so it's not just for initial placement. The canonical link is the rigid body that the `__model__` frame is attached to since all frames, implicit or explicit, must terminate on a rigid body.

There's a debugging tool that visualizes the frame graph. I think that might help to clarify.

```
xdot <(ign sdf -g frame test/sdf/model_frame_attached_to.sdf)
```

File: sdf/1.8/pose.sdf (hunk 9 above)
Comment (azeey) [reply sdf/1.8/pose.sdf]: Yes, that's why I said "technically". When the file is nested via `//include`, it becomes part of the current file, and references into that nested model are allowed. The encapsulation is alluding to the fact that the current file can't reference a frame that exists in another file that's going to include the current file. i.e., if model `A` includes model `B`,  you can have references like `B::some_frame` inside model `A`, but you can't have `A::some_frame` inside model `B`.

Comment (FirefoxMetzger) [reply sdf/1.8/pose.sdf]: In this case, I would say it depends on what your vision for this feature is. If there are plans to extend SDFormat and make referencing frames inside an `//include` element an explicit feature it does make sense to remove the "same file" reference.  (Maybe it is already a feature because it is possible to interpret the proposal in this way, too.)

If it is meant to remain an implementation detail and users shouldn't rely on this behavior I would - personally - keep saying same file here. It is probably better to document the spec/API in the user-facing docs instead of documenting the implementation.

File: sdf/1.7/pose.sdf (hunk 6 above)
Comment (FirefoxMetzger) [reply sdf/1.7/pose.sdf]: I updated the description to 

```
 If missing, the pose is expressed in the frame of the parent element. The
parent element is the grandparent of this pose element, i.e., the parent of
the element that has its pose set by this pose element.
```

I think the latter part is a bit cryptic and sounds artificial; I didn't manage to come up with nicer wording. Suggestions for improvement are welcome :)

**Comment by FirefoxMetzger** [comment]
@azeey Once the tests pass, I'm happy with this PR unless you have any further suggestions/comments.

File: sdf/1.7/pose.sdf (hunk 6 above)
Comment (azeey) [reply sdf/1.7/pose.sdf]: How about
>If missing, the pose is expressed in the frame of the parent XML element of the element that contains the pose. For exceptions to this rule and more details on the default behavior, see http://sdformat.org/tutorials?tut=pose_frame_semantics&ver=1.5 and http://sdformat.org/tutorials?tut=pose_frame_semantics&ver=1.7

The exceptions are:
* `//joint/pose`: the default pose is relative to the child link/frame
* `//frame/pose`: the default pose is relative to the attached_to frame.

Also, I think just saying parent element can be confusing for joints, which have parents and children, so I changed it to "parent XML element".

File: sdf/1.7/frame.sdf [hunk 10]
```diff
@@ -1,16 +1,32 @@
...
+      If missing, this frame is attached to the containing scope's frame. Within
+      a //world scope this is the implicit world frame, and within a //model
+      scope this is the implicit model frame.
+
+      A frame moves jointly with the frame it is @attached_to. This is different
+      from //pose/@relative_to. @attached_to defines how the frame is attached
+      to a //link, //model, or //world frame, while //pose/@relative_to defines
+      how the frame's pose is represented numerically. As a result, following
+      the chain of @attached_to attributes must always lead to a //link,
+      //model, //world, or //joint (implicitly attached_to its child //link).
```
Comment (azeey) [inline sdf/1.7/frame.sdf:29]: I don't see the value of adding `//joint` here because that's not where the chain terminates. It terminates on the child link, which is covered by the previous portion of the sentence.

File: sdf/1.8/frame.sdf [hunk 11]
```diff
@@ -1,16 +1,32 @@
...
+      If missing, this frame is attached to the containing scope's frame. Within
+      a //world scope this is the implicit world frame, and within a //model
+      scope this is the implicit model frame.
+
+      A frame moves jointly with the frame it is @attached_to. This is different
+      from //pose/@relative_to. @attached_to defines how the frame is attached
+      to a //link, //model, or //world frame, while //pose/@relative_to defines
+      how the frame's pose is represented numerically. As a result, following
+      the chain of @attached_to attributes must always lead to a //link,
+      //model, //world, or //joint (implicitly attached_to its child //link).
```
Comment (azeey) [inline sdf/1.8/frame.sdf:29]: For 1.8, the chain doesn't terminate on a model, since we now have visibility into a nested model. In SDFormat 1.7, nested models weren't fully supported, so they were treated as opaque objects and the attached_to chain was allowed to terminate on a nested model. So, now we can just say "...must always lead to a //link or //world".

File: sdf/1.8/pose.sdf [hunk 12]
```diff
@@ -1,10 +1,24 @@
...
-      Name of frame relative to which the pose is applied.
+      If specified, this pose is expressed in the named frame. The named frame
+      must be declared within the same scope (world/model) as the element that
+      has its pose specified by this tag.
+
+      If missing, the pose is expressed in the frame of the parent element. The
+      parent element is the grandparent of this pose element, i.e., the parent of
+      the element that has its pose set by this pose element.
+
+      Note that @relative_to merely affects an element's initial position and
```
Comment (azeey) [inline sdf/1.8/pose.sdf:16]:
```suggestion
      Note that @relative_to merely affects an element's initial pose and
```

**Review by azeey — COMMENTED** [review]
Just a couple more tweaks. Looks great!

File: sdf/1.7/pose.sdf (hunk 6 above)
Comment (FirefoxMetzger) [reply sdf/1.7/pose.sdf]: Sounds very nice; I've updated the description.

I'm actually a bit back and forth on adding links. I really like the cross-referencing of the spec and the docs; when I first got into SDFormat I didn't realize that documentation exists to the extent it does and I wouldn't have known where to look. 

At the same time, the current spec website doesn't allow clicking hyperlinks in descriptions (because the entire element is a hyperlink to the element's fragment, see for example: http://sdformat.org/spec?ver=1.8&elem=sensor#camera_distortion). As a result, I am using the chrome debug tool to extract links from the spec which is sub-optimal. Perhaps something low-priority to look into in the future :)

File: sdf/1.7/frame.sdf (hunk 10 above)
Comment (FirefoxMetzger) [reply sdf/1.7/frame.sdf:29]: This was added during the review by @scpeters : https://github.com/ignitionrobotics/sdformat/pull/666#discussion_r690636440

It adds value in the sense that `//joint` doesn't have a @attached_to attribute and thus you can't follow `//joint/@attached_to` to the child `//link` - strictly speaking. If we are really trying to be picky, I guess we could also remove `//model`, because it implicitly attaches to a link via `//model/@canonical_link`.

Personally, I'm happy either way as long as it is clear that any chain must terminate in a `//link` or `//world`.

File: sdf/1.8/frame.sdf (hunk 11 above)
Comment (FirefoxMetzger) [reply sdf/1.8/frame.sdf:29]: Ha! Great minds think alike I guess. This seems to be similar to the point above. I'll fix it in the same way once there is consensus.

File: sdf/1.8/pose.sdf (hunk 12 above)
Comment (FirefoxMetzger) [reply sdf/1.8/pose.sdf]: fixed. I can't directly apply it from GH, because this breaks the DCO test (wrong username when signing off).

File: sdf/1.7/frame.sdf (hunk 10 above)
Comment (azeey) [reply sdf/1.7/frame.sdf:29]: Ah, I see this is specifically following `@attached_to` attributes, not necessarily the frame attachments in the frame graph. Okay, we can leave `//joint` here.

**Review by azeey — APPROVED** [review]

File: sdf/1.7/pose.sdf [hunk 13]
```diff
@@ -1,10 +1,22 @@
...
-      Name of frame relative to which the pose is applied.
+      If specified, this pose is expressed in the named frame. The named frame
+      must be declared within the same scope (world/model) as the element that
+      has its pose specified by this tag.
+
+      If missing, the pose is expressed in the frame of the parent XML element
+      of the element that contains the pose. For exceptions to this rule and
+      more details on the default behavior, see
+      http://sdformat.org/tutorials?tut=pose_frame_semantics and
+      http://sdformat.org/tutorials?tut=pose_frame_semantics
```
Comment (scpeters) [inline sdf/1.7/pose.sdf:16]: these links are now identical

Comment (azeey) [reply sdf/1.7/pose.sdf:16]: Oops, they were supposed to be pointing to the 1.5 and 1.7 versions respectively.

Comment (FirefoxMetzger) [reply sdf/1.7/pose.sdf:16]: Yep, you pasted versioned linkes @azeey , but the CI chain is unhappy about the fragment character (`#`) in the link and breaks if it is there. I removed them but didn't realize that it makes the link the same. 

We could patch it and remove the second link.

Comment (FirefoxMetzger) [reply sdf/1.7/pose.sdf:16]: See: https://github.com/ignitionrobotics/sdformat/pull/702

**Comment by osrf-triage** [comment]
This pull request has been mentioned on **Gazebo Community**. There might be relevant details there:

https://community.gazebosim.org/t/new-ignition-releases-2022-03-25-fortress-edifice-citadel/1343/1
